homeassistant/components/asuswrt/router.py

```diff
@@ -46,41 +46,84 @@
 )
 
 CONF_REQ_RELOAD = [CONF_DNSMASQ, CONF_INTERFACE, CONF_REQUIRE_IP]
+
+KEY_COORDINATOR = "coordinator"
+KEY_SENSORS = "sensors"
+
 SCAN_INTERVAL = timedelta(seconds=30)
 
+SENSORS_TYPE_BYTES = "sensors_bytes"
+SENSORS_TYPE_COUNT = "sensors_count"
+SENSORS_TYPE_RATES = "sensors_rates"
+
 _LOGGER = logging.getLogger(__name__)
 
 
-class AsusWrtSensorType(enum.IntEnum):
-    """Enun possible AsusWrt sensor types."""
+class AsusWrtSensorDataHandler:
+    """Data handler for AsusWrt sensor."""
 
-    TypeNone = 0
-    TypeCount = 1
-    TypeBytes = 2
-    TypeRates = 3
+    def __init__(self, hass, api):
+        """Initialize a AsusWrt sensor data handler."""
+        self._hass = hass
+        self._api = api
+        self._connected_devices = 0
 
+    async def _get_connected_devices(self):
+        """Return number of connected devices."""
+        return {SENSOR_CONNECTED_DEVICE: self._connected_devices}
 
-class AsusWrtSensorData:
-    """Representation of AsusWrt sensor data."""
+    async def _get_bytes(self):
+        """Fetch byte information from the router."""
+        ret_dict: Dict[str, Any] = {}
+        try:
+            datas = await self._api.async_get_bytes_total()
+        except OSError as ex:
+            raise UpdateFailed(ex)
 
-    def __init__(self, sensor_type: AsusWrtSensorType, value=None):
-        """Initialize a AsusWrt sensor data."""
-        self.type = sensor_type
-        self.value = value
-        self._enabled = False
+        ret_dict[SENSOR_RX_BYTES] = datas[0]
+        ret_dict[SENSOR_TX_BYTES] = datas[1]
 
-    def enable(self):
-        """Set sensor enabled."""
-        self._enabled = True
+        return ret_dict
 
-    def disable(self):
-        """Set sensor disabled."""
-        self._enabled = False
+    async def _get_rates(self):
+        """Fetch rates information from the router."""
+        ret_dict: Dict[str, Any] = {}
+        try:
+            rates = await self._api.async_get_current_transfer_rates()
+        except OSError as ex:
+            raise UpdateFailed(ex)
+
+        ret_dict[SENSOR_RX_RATES] = rates[0]
+        ret_dict[SENSOR_TX_RATES] = rates[1]
+
+        return ret_dict
+
+    def update_device_count(self, conn_devices):
+        """Update connected devices attribute."""
+        self._connected_devices = conn_devices
+
+    async def get_coordinator(self, sensor_type: str):
+        """Get the coordinator for a specific sensor type."""
+        if sensor_type == SENSORS_TYPE_COUNT:
+            method = self._get_connected_devices
+        elif sensor_type == SENSORS_TYPE_BYTES:
+            method = self._get_bytes
+        elif sensor_type == SENSORS_TYPE_RATES:
+            method = self._get_rates
+        else:
+            return None
```

Comment:
Raise a RuntimeError instead or similar. This should just be a developer mistake and not something we need to handle downstream.